### PR TITLE
Common Subexpression Elimination

### PR DESCRIPTION
**Pull request by isuruf**

### CODE REVIEW

**Comment by certik** [comment]
Nice!!

File: symengine/basic-inl.h [hunk 1]
```diff
@@ -103,6 +103,15 @@ hash_t vec_hash<T>::operator()(const T &v) const
...
 
+//! workaround for MinGW bug
+template <typename T>
+std::string to_string(const T &value)
+{
+    std::ostringstream ss;
+    ss << value;
+    return ss.str();
+}
+
```
Comment (certik) [inline symengine/basic-inl.h]: Is the bug that `std::to_string` is not available in MinGW? For now this is good enough, later perhaps we can somehow inject it into the `std` namespace, just for MinGW, perhaps cmake can figure this out.

Comment (isuruf) [reply symengine/basic-inl.h]: Yes. `std::to_string` is not available on MinGW.

**Comment by isuruf** [comment]
Sourceforge is experiencing issues. https://twitter.com/sfnet_ops/status/912673956712341514

**Comment by certik** [comment]
I saw that. Besides that, is this PR finished?

**Comment by isuruf** [comment]
> Besides that, is this PR finished?

Yes

**Comment by isuruf** [comment]
For `large_exprs.TimeLargeExpressionOperations.time_cse` in sympy benchmark suite, sympy takes `65.88 ms` while symengine takes `3.76 ms`

**Pull request merged**